### ISSUE TITLE
Yocto Layer

### ISSUE THREAD

**Issue by markrey** [calamares/calamares#919]
#### Submission type

  - [ ] Bug report
  - [X ] Feature Request

#### Info regarding which version of Calamares is used, which Distribution

> Latest

We are using Yocto to build our image.
Wanted to check if Yocto layer support is something under planning. 

Thanks on the great open-source product!

**Comment by adriaandegroot** [comment]
I don't use Yocto or know how it works, so you'll have to explain: what are Yocto layers, how do they work, and what would "supporting" Yocto layers mean.

None of the distro's that use Calamares seem to be Yocto-based, so a short answer would be "no".

**Comment by philmmanjaro** [comment]
Normally it should be possible to support [Yocto](https://www.yoctoproject.org/software-overview/). Is there an example ISO for reviewing the current structure. Also which installer is now used?

**Comment by kkofler** [comment]
How do you want to support Yocto? Yocto targets embedded devices, which do not use the live images that are normally the installation source for Calamares. Instead, it produces a file system image, which is directly copied to the target device using a very device-specific method. (If you're lucky, it's enough to dd it to an SD card and insert that into the device, but others are more complicated.) There isn't really a good place to bring up an installer like Calamares in that process.

What could make sense to use in that setup is the no-chroot (OEM) mode, where Calamares runs on the already installed image, allowing the user to set up user settings, but skipping the partitioning (`partition`) and installing (`unpackfs`) steps.

**Comment by adriaandegroot** [comment]
@markrey This needs explanation from **you**, because we don't even understand what the question means. You're welcome to drop in on `#calamares` on FreeNode IRC for more explanation.

**Comment by philmmanjaro** [comment]
@markrey: so you want to replace **Anaconda** with **Calamares** on your distro. For **Manjaro** we are using [scripts](https://raw.githubusercontent.com/manjaro/packages-extra/master/calamares/PKGBUILD) to package **Calamares** for **Pacman**. So someone needs to create **recipes** or **layers** to make this work for **Yocto**. Even I looked it up a little, I'm currently still lost in how to even accomplish that ...

**Comment by adriaandegroot** [comment]
Mostly what @philmmanjaro said: you can package up Calamares as you like (I see .deb and .rpm in that diagram of yours) so you can probably borrow existing build recipes (e.g. from Debian / siduction / KDE Neon / KaOS) to build the .deb. Then make sure it is on your live image. An installed Calamares usually has a `~liveuser/Desktop/Calamares.desktop` file so that they can run it from their live X session. There are two modes for Calamares: regular, where you install from the live system onto a storage device (e.g. hard disk) with partitioning and bootloader installation, and *OEM*, where you're working inside an already-installed system image (e.g. on an SD card or eMMC for a SBC) which is mostly about first-run configuration.

**Comment by kkofler** [comment]
You may also want to see the Fedora packaging of Calamares, which is obviously the most compatible with Anaconda.

**Comment by abucodonosor** [comment]
@philmmanjaro  @adriaandegroot 

I looked at these and that is not an Calamares task.

Also packaging seems to not be that hard, there are files called xxx.bb which
seems to have a simple syntax and some sort gentoo like schemas support
for all kind things , however that is not an taks for Calamares developrs

Also rpm , deb is the output of .bb files not the input.. 

@markrey 

*Someone* from Yocto (!?) would need to package Qt5 and kf5 parts.
None of these exists .. there is Qt3 ( OMG ) and some Qt4 and that is.

I also checked openembedded and they don't have Qt* at all.

So in short there is nothing to support for Calamares not even 
after all dependecies exists since is you who need to create an calamares.bb
for your Distribution with right configuration files ( an I guess some custom 
startup scripts ) and so on ..

**Comment by abucodonosor** [comment]
@kkofler 

Would be probably nice but how I read all the Yocto things there is no way of re-using
an rpm.. or anything else but .bb files

**Comment by markrey** [comment]
Thank you very much Everyone!
I think i got a good handle on how it should be scripted.

The only complexity is bundling with respect to multiple types of end releases. 
- Some devices do not have "Display" capabilities so (like some ARM devices) 
- As well as various network layers for installation.

@adriaandegroot 
I agree on the approach. It will be much simpler that way. And both Calamares and Final distros are not deeply integrated / properly isolated.

@abucodonosor 
Someone from Yocto (!?) would need to package Qt5 and kf5 parts.

Yepp i agree. But it should not be a big task for us. We can do it quite easily. 

@abucodonosor @kkofler 
There is a way to use RPM / DEB as a part of installation as you can just write is in basic layer file.

I am closing the ticket for now as it is more of an extension to Calamares rather than Calamares itself. 
I will reopen with specific questions if required.

Once successful, i will also create PULL request with documentation / script.
I really appreciate everyone's support!

**Comment by adriaandegroot** [comment]
> Some devices do not have "Display" capabilities so (like some ARM devices)

Without a display, an interactive, X11-based, user-driven installation tool is not going to be very useful.

> Someone from Yocto (!?) would need to package Qt5 and kf5 parts.

Johan Thelin and Volker Krause do just that, actually.
